exec/java-exec/src/main/java/org/apache/drill/exec/work/batch/UnlimitedRawBatchBuffer.java

```diff
@@ -90,14 +100,39 @@ public boolean isEmpty() {
 
     @Override
     public void add(RawFragmentBatch batch) {
+      int recordCount = batch.getHeader().getDef().getRecordCount();
+      long bathByteSize = batch.getByteCount();
+      if (recordCount != 0) {
+        //skip first header batch
+        totalBatchSize += bathByteSize;
+        sampleTimes++;
+      }
+      if (sampleTimes == maxSampleTimes) {
+        long averageBathSize = totalBatchSize / sampleTimes;
+        //make a decision
+        long limit = context.getAllocator().getLimit();
+        long thresholdNetworkMem = (long) (limit * 0.40);
+        if (averageBathSize > 0) {
+          runtimeSoftLimit = (int) (thresholdNetworkMem / averageBathSize);
+          runtimeAckCredit = runtimeSoftLimit / fragmentCount;
+          runtimeAckCredit = runtimeAckCredit > 0 ? runtimeAckCredit : 1;
+        }
+      }
+      if (runtimeSoftLimit > 0) {
```

Comment:
As discussed above, `runtimeSoftLimit` can be 0 even if we do the calcs under certain conditions. Maybe check `runtimeAckCredit` instead, which we have forced to be non-zero?

Reply:
If we get a 0 runtimeSoftLimit, we do nothing just to keep the original static credit flow control.